### ISSUE TITLE
Keypoint colors do not match with classes

### ISSUE THREAD

**Issue by d-kleine** [ultralytics/ultralytics#14389]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Val, Predict

### Bug

Using Ultralytics version: 8.2.54

When using yolov8n-pose for pose estimation, it seems like keypoint colors do not match with with the classes. I have two classes, each class should have its own color, but both in the training as well in the prediction, the keypoints have more than two different colors (even though it's only two classes).

https://github.com/d-kleine/barrel_detection/tree/a1a4480ac2cdc01e49876b9dca49b70a70ad59af/runs/barrel/yolov8n_custom

Example:
* Ground truth
![val_batch0_labels](https://github.com/user-attachments/assets/a58dfece-b66f-44ee-8298-3edba4d7eb54)

* Prediction
![val_batch0_pred](https://github.com/user-attachments/assets/d0590ac2-5a73-47f1-9e3d-7c0c25b21678)

Actually the colors of the keypoint should match with class, for example like this (this is how I labeled my data in Roboflow):
![grafik](https://github.com/user-attachments/assets/66be7ed1-4192-4579-9d99-1dca40562f84)
![grafik](https://github.com/user-attachments/assets/33cc7e40-19a7-45d9-b994-1bc9e860ad38)

Roboflow project: https://universe.roboflow.com/barrellocalization/barrel_pose

### Minimal Reproducible Example

```python
from ultralytics import YOLO

model = YOLO("yolov8n-pose.pt")
 
# Training.
results = model.train(
   data='data.yaml',
   imgsz=640,
   epochs=10,
   name='yolov8n_custom',
   val=True,
   project='./runs/barrel',
   seed=42,
   exist_ok=True,
   )
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by d-kleine** [comment]
Along with this bug report, I would also request a feature: I think it would be also useful to see the keypoint IDs next to the keypoints. For example, image you want to do pose estimation for a cube, how would you see from the generated bbox+kpt plot if the prediction is correct?
 
![cubes_kpt_ids](https://github.com/user-attachments/assets/c6cbbb11-ff4c-4067-81f9-73ad0e433a71)

Currently, there are no keypoints IDs (or keypoint labels) displayed:
![grafik](https://github.com/user-attachments/assets/1e87b5b7-df8d-4419-8279-21954e3f77b7)

**Comment by glenn-jocher** [comment]
@d-kleine thank you for bringing this to our attention! 

Regarding the keypoint color mismatch, it seems like there might be an issue with how the keypoints are being visualized. To help us diagnose this, could you please ensure that you are using the latest version of the Ultralytics package? If the issue persists, providing a minimal reproducible example would be very helpful. You can refer to our [minimum reproducible example guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details.

As for your feature request to display keypoint IDs next to the keypoints, this is a great suggestion! It would indeed be beneficial for better visual verification of pose estimation results. We will consider this for future updates.

If you have any further details or examples, please share them. Your feedback is invaluable in helping us improve our models and tools. Thank you! 🚀

**Comment by glenn-jocher** [comment]
Thank you for confirming the issue with the latest version and providing a reproducible example. This is very helpful! 

We will investigate the keypoint color mismatch further. In the meantime, if you have any additional insights or observations, please feel free to share them.

Regarding your feature request to display keypoint IDs next to the keypoints, we appreciate the suggestion and will consider it for future updates. 

Thank you for your patience and for contributing to improving our models and tools! 🙌

**Comment by d-kleine** [comment]
Linking feature request #14404 as it is related to the keypoints

**Comment by glenn-jocher** [comment]
Thank you for linking the related feature request #14404. We appreciate your thoroughness! 

We are actively investigating the keypoint color mismatch issue. If you have any additional details or observations, please share them here.

Regarding the feature request to display keypoint IDs, we see its potential value and will consider it for future updates. Your feedback is invaluable in helping us improve our models and tools. Thank you for your contributions! 🙌

**Comment by glenn-jocher** [comment]
@d-kleine I looked into this a bit more. The Annotator supplies a separate keypoint map here that you can adjust to your custom requirements. Keypoints are supposed to be colored differently, otherwise no one would be able to determine which keypoints where which, and then we'd be getting bug reports stating that all keypoints are impossible to tell apart.

https://github.com/ultralytics/ultralytics/blob/21ca23568105848901439d9ebcb62d237a53adb1/ultralytics/utils/plotting.py#L34-L186

**Comment by d-kleine** [comment]
> Keypoints are supposed to be colored differently

Maybe a language barrier here. I was not saying that the keypoint should not have different color. 

See this example, this is the ground truth: How would you see from this prediction which keypoints belong to what class?
![val_batch0_labels](https://github.com/user-attachments/assets/a58dfece-b66f-44ee-8298-3edba4d7eb54)

When comparing it with the annotated figure from Roboflow, you can see that the "0_turret" class has 3 differently colored keypoints inside (barrel's end =dark blue; barrel's start = cyan, top hatch = cyan, bottom = white):
![grafik](https://github.com/user-attachments/assets/33cc7e40-19a7-45d9-b994-1bc9e860ad38)

-> **This does not make sense at all!** <-

My suggestion was to every keypoints corresponding to a specific class found an image should have the same color. So, for example, you have an image with two different objects in the an image. The first object is "object A", the second one is "object B". "object A" will get a **blue** bounding box at it gets assigned to class A. So, it should get also a **blue** color for the keypoints. "object B" will get an **orange** bounding box at it gets assigned to class B. So, it should get also an **orange** color for the keypoints.
![grafik](https://github.com/user-attachments/assets/35b5accc-90d0-449e-9495-2c609f5170ce)

Currently, there are more colors displayed than classes. In my above image, I have two classes, but there 3 colors (which, cyan, and dark blue) were created. I think this is bad usability because I cannot understand which keypoint belongs to which class.

If at least the color of the bounding box (that matches with the class) would maps the same color to its keypoints, it would be easier to understand.

> otherwise no one would be able to determine which keypoints where which, and then we'd be getting bug reports stating that all keypoints are impossible to tell apart.

That is currently my issue - having more colors than classes (which does not make sense at all), no keypoint ID labels, no keypoints name labels). That's why I have opened two feature requests on that. This one is the bug (at least from my pov), but having the option to display keypoint ID labels and/or keypoint name labels in prediction won't make life and usage easier.

**Comment by pderrenger** [comment]
Thank you for the clarification. I understand your concern now. The issue seems to be that the keypoint colors do not consistently match the class colors, making it difficult to distinguish which keypoints belong to which class.

Your suggestion to have keypoints colored according to their class makes sense for better visual clarity. We will look into this to improve the usability. 

For now, you might consider customizing the keypoint colors in the Annotator as a temporary workaround. 

Thank you for your valuable feedback!

**Comment by glenn-jocher** [comment]
@d-kleine here's an example of the standard keypoint coloring, which is color by point rather than color by instance or color by class. This allows users to identify if an individual keypoint is correctly placed, which is the most common use case.
https://www.youtube.com/watch?v=Y28xXQmju64&t=133s

<img width="1796" alt="Screenshot 2024-07-14 at 14 55 33" src="https://github.com/user-attachments/assets/7f87b287-f8b7-406b-8f5d-a068ec14d901">

**Comment by d-kleine** [comment]
This also colors the connection lines between the keypoints, not the keypoints themselves. At it's some kind of instance coloring because the connection lines as well as the keypoints have the same color for an instance (e.g. head, torso, etc.)..

I also think that he uses `model.predict()` for running the inference (because of the `conf` param being used). And there is only one class, image there would be 2 or more classes, this would be a visual clutter.

The random keypoint coloring in the output of `model.train()` still does not make any sense:
![val_batch0_labels](https://github.com/user-attachments/assets/a58dfece-b66f-44ee-8298-3edba4d7eb54)

All I am asking for is a better way for a prediction to see which keypoints belong to what, whether an a specific class, specific instances. My suggestion would be to color keypoints in the output `model.train()` by class by default. And add a param to steer this behavior, for example:

```python
model.train(
    ...
    keypoint_color_mode='class'
```

With `model.predict()`, you could give more customization options:

```python
model.predict(
    ...
    kpt_color_mode='class',  # or 'instance'
    kpt_draw_connections=True  # True for drawing connection line between closest keypoints
)
```

This is just an idea, that needs further investigation. But please make the output image of both GT as well as Prediction visually intuitive and better to understand.

**Comment by pderrenger** [comment]
Thank you for your detailed feedback. We understand the importance of having clear and intuitive visualizations for keypoints, especially when dealing with multiple classes or instances. Your suggestions for customizable keypoint coloring modes and connection lines are valuable and could indeed enhance the usability of the tool.

For now, please ensure you're using the latest version of the Ultralytics package, as updates often include improvements and bug fixes. We will consider your suggestions for future updates to make the output more visually intuitive. If you have any further insights or suggestions, feel free to share them. Your input is greatly appreciated!
